### PR TITLE
Always report object states in same order, regardless whether positive or negative

### LINKED ISSUE #7076: Inconsistent reporting of checkable list items
Can be reproduced in many lists with checkable items in Windows 10, e.g. the disk clean-up utility.
Uncecked items are brailled as `Foo sel ( ) chk`, checked ones as `Foo (x) sel chk`.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
fixes #7076

### Summary of the issue:
When reporting object states to the user in speech or braille, the positive object states are processed before the negative states. This creates the following problem as reported by @dkager:

> Can be reproduced in many lists with checkable items in Windows 10, e.g. the disk clean-up utility.
> Unchecked items are brailled as  Foo sel ( ) chk , checked ones as  Foo (x) sel chk .

### Description of how this pull request fixes the issue:
1. For both speech and braille, de positive and negative states sets are merged and numerically sorted. When looping through this merged, sorted list of states, we check whether the state is positive or negative and the appropriate label will be spoken or brailled.
2. The negative state for controlTypes.STATE_DROPTARGET (Done dragging) is moved from speech.py to controlTypes.py

### Testing performed:
In disk clean up, the selected state is now always spoken before the checked state, regardless whether positive or negative

### Known issues with pull request:
1. As I noted in https://github.com/nvaccess/nvda/issues/7076#issuecomment-307720826, This changes the order of presentation people are accustomed to in some situations. E.g. invalid entry required will change to required invalid entry. Do we care about this, or do we want to have a certain order of priority as @JCSTeh suggested in https://github.com/nvaccess/nvda/issues/7076#issuecomment-305350599?
2. We probably want to test whether the "done dragging" case still works correctly.

### CODE REVIEW

File: source/braille.py [hunk 1]
```diff
@@ -618,9 +618,17 @@ def getBrailleTextForProperties(**propertyValues):
...
-		textList.extend(positiveStateLabels.get(state, controlTypes.stateLabels[state]) for state in positiveStates)
 		negativeStates = controlTypes.processNegativeStates(role, states, controlTypes.REASON_FOCUS, None)
-		textList.extend(negativeStateLabels.get(state, controlTypes.negativeStateLabels.get(state, _("not %s") % controlTypes.stateLabels[state])) for state in negativeStates)
+		for state in sorted(positiveStates | negativeStates):
+			if state in positiveStates:
+				textList.append(positiveStateLabels.get(state, controlTypes.stateLabels[state]))
+			elif state in negativeStates:
+				# Translators: Indicates that a particular state on an object is negated.
+				# Separate braille abbreviations have now been defined for commonly negated states (e.g. ( ) for not checked),
+				# 			and, if no braille abbreviations are provided, we fall back to the defined state labels in ControlTypes,
```
Comment (dkager) [inline source/braille.py]: Slightly more spaces than necessary. :)

**Review by dkager — COMMENTED** [review]
Looks good to me. Quite a bit of code duplicaion, but I guess the states for braille and speech are not guaranteed to be identical?

File: source/speech.py [hunk 2]
```diff
@@ -937,26 +937,19 @@ def getSpeechTextForProperties(reason=controlTypes.REASON_QUERY,**propertyValues
...
 	states=propertyValues.get('states')
 	realStates=propertyValues.get('_states',states)
-	if states is not None:
-		positiveStates=controlTypes.processPositiveStates(role,realStates,reason,states)
-		textList.extend([controlTypes.stateLabels[x] for x in positiveStates])
-	if 'negativeStates' in propertyValues:
-		negativeStates=propertyValues['negativeStates']
-	else:
-		negativeStates=None
-	if negativeStates is not None or (reason != controlTypes.REASON_CHANGE and states is not None):
```
Comment (leonardder) [inline source/speech.py]: @JCSTeh: Could you clarify this line? Was there a particular reason why if negativeStates is not None wasn't enough?

File: source/controlTypes.py [hunk 3]
```diff
@@ -711,3 +714,18 @@ def processNegativeStates(role, states, reason, negativeStates):
 		# This is not a state change; only positive states were supplied.
 		# Return all negative states which should be spoken, excluding the positive states.
 		return speakNegatives - states
+
+def processAndLabelStates(role, states, reason, positiveStates, negativeStates, positiveStateLabelDict={}, negativeStateLabelDict={}):
```
Comment (leonardder) [inline source/controlTypes.py]: I'm happy to provide a Doc-string for this, but processPositiveStates and processNegativeStates don't have a doc string yet. May be the initial author of these functions can provide some explanation (e.g. about why we both have states and positiveStates parameters for processPositiveStates)? We can add doc strings for these in this pr as well.

File: source/speech.py (hunk 2 above)
Comment (jcsteh) [inline source/speech.py]: > Could you clarify this line? Was there a particular reason why if negativeStates is not None wasn't enough?

The `negativeStates` argument is only provided for `REASON_CHANGE`. It specifies states that have just changed from positive to negative. However, for other reasons (e.g. `REASON_FOCUS`), we still want to report certain states if they aren't present; e.g. "not checked".

File: source/controlTypes.py (hunk 3 above)
Comment (jcsteh) [inline source/controlTypes.py]: > why we both have states and positiveStates parameters for processPositiveStates

`states` contains the current states for the object. For REASON_CHANGE, `positiveStates` contains only those states that just got added. In other words, `positiveStates` contains only the states we want to possibly be reported at this time, but we might need the "unfiltered" states for the object, so we provide `states` as well.

**Comment by leonardder** [comment]
@JCSTeh: Apart from doc strings, I think this is ready for review.

File: source/speech.py (hunk 2 above)
Comment (leonardder) [inline source/speech.py]: Thanks. In the new code, it seems this check is no longer necessary as I provide empty sets as default

**Comment by leonardder** [comment]
This now contains doc strings as well. @Dkager: Would you like to review them? Your grammar is better than mine. :)

File: source/braille.py [hunk 4]
```diff
@@ -476,10 +476,7 @@ def getBrailleTextForProperties(**propertyValues):
 	if value and role not in controlTypes.silentValuesForRoles:
 		textList.append(value)
 	if states:
-		positiveStates = controlTypes.processPositiveStates(role, states, controlTypes.REASON_FOCUS, states)
-		textList.extend(positiveStateLabels.get(state, controlTypes.stateLabels[state]) for state in positiveStates)
-		negativeStates = controlTypes.processNegativeStates(role, states, controlTypes.REASON_FOCUS, None)
-		textList.extend(negativeStateLabels.get(state, controlTypes.negativeStateLabels.get(state, _("not %s") % controlTypes.stateLabels[state])) for state in negativeStates)
+		textList.extend(controlTypes.processAndLabelStates(role, states, controlTypes.REASON_FOCUS, states,None,positiveStateLabels,negativeStateLabels))
```
Comment (dkager) [inline source/braille.py]: Inconsistent use of space after comma. Maybe pull this function out into two lines to improve readability?

File: source/controlTypes.py [hunk 5]
```diff
@@ -573,6 +573,9 @@
 	STATE_PRESSED:_("not pressed"),
 	# Translators: This is presented when a checkbox is not checked.
 	STATE_CHECKED:_("not checked"),
+	# Translators: This is presented when drag and drop is finished.
+	# This is only reported for objects which support accessible drag and drop.
```
Comment (dkager) [inline source/controlTypes.py]: I think "that support" is the correct form here.

File: source/controlTypes.py [hunk 6]
```diff
@@ -638,6 +641,20 @@
 }
 
 def processPositiveStates(role, states, reason, positiveStates):
+	"""Processes the states for an object and returns the positive states to output for a specified reason.
+	For example, if C{STATE_CHECKED} is among the returned states, it means that the processed object is checked.
```
Comment (dkager) [inline source/controlTypes.py]: Trivial I guess, but "in the returned states" is a bit closer to what you'd actually write in a program to check this.

File: source/controlTypes.py [hunk 7]
```diff
@@ -638,6 +641,20 @@
 }
 
 def processPositiveStates(role, states, reason, positiveStates):
+	"""Processes the states for an object and returns the positive states to output for a specified reason.
+	For example, if C{STATE_CHECKED} is among the returned states, it means that the processed object is checked.
+	@param role: The role of the object to process states for (e.g. C{ROLE_CHECKBOX}.
+	@type role: int
+	@param states: Usually the raw states for an object to process.
```
Comment (dkager) [inline source/controlTypes.py]: This implies there are exceptions. What are they?

File: source/controlTypes.py [hunk 8]
```diff
@@ -638,6 +641,20 @@
 }
 
 def processPositiveStates(role, states, reason, positiveStates):
+	"""Processes the states for an object and returns the positive states to output for a specified reason.
+	For example, if C{STATE_CHECKED} is among the returned states, it means that the processed object is checked.
+	@param role: The role of the object to process states for (e.g. C{ROLE_CHECKBOX}.
+	@type role: int
+	@param states: Usually the raw states for an object to process.
+	@type states: set
+	@param reason: The reason to process the states for (e.g. C{REASON_FOCUS}.
```
Comment (dkager) [inline source/controlTypes.py]: I think "for" should be removed.

File: source/controlTypes.py [hunk 9]
```diff
@@ -683,6 +700,20 @@ def processPositiveStates(role, states, reason, positiveStates):
 	return positiveStates
 
 def processNegativeStates(role, states, reason, negativeStates):
+	"""Processes the states for an object and returns the negative states to output for a specified reason.
+	For example, if C{STATE_CHECKED} is among the returned states, it means that the processed object is not checked.
```
Comment (dkager) [inline source/controlTypes.py]: Same stuff as above.

File: source/controlTypes.py [hunk 10]
```diff
@@ -711,3 +742,38 @@ def processNegativeStates(role, states, reason, negativeStates):
 		# This is not a state change; only positive states were supplied.
 		# Return all negative states which should be spoken, excluding the positive states.
 		return speakNegatives - states
+
+def processAndLabelStates(role, states, reason, positiveStates, negativeStates, positiveStateLabelDict={}, negativeStateLabelDict={}):
```
Comment (dkager) [inline source/controlTypes.py]: In another PR @jcsteh cautioned against using empty dicts as default function parameters. I wonder if that applies here.

File: source/controlTypes.py [hunk 11]
```diff
@@ -711,3 +742,38 @@ def processNegativeStates(role, states, reason, negativeStates):
 		# This is not a state change; only positive states were supplied.
 		# Return all negative states which should be spoken, excluding the positive states.
 		return speakNegatives - states
+
+def processAndLabelStates(role, states, reason, positiveStates, negativeStates, positiveStateLabelDict={}, negativeStateLabelDict={}):
+	"""Processes the states for an object and returns the appropriate state labels for both positive and negative states.
+	@param role: The role of the object to process states for (e.g. C{ROLE_CHECKBOX}.
+	@type role: int
+	@param states: Usually the raw states for an object to process.
```
Comment (dkager) [inline source/controlTypes.py]: Same stuff as above.

File: source/controlTypes.py [hunk 12]
```diff
@@ -711,3 +742,38 @@ def processNegativeStates(role, states, reason, negativeStates):
...
+	@rtype: [str, ...]
+	"""
+	mergedStateLabels=[]
+	positiveStates = processPositiveStates(role, states, reason, positiveStates)
+	negativeStates = processNegativeStates(role, states, reason, negativeStates)
+	for state in sorted(positiveStates | negativeStates):
+		if state in positiveStates:
+			mergedStateLabels.append(positiveStateLabelDict.get(state, stateLabels[state]))
+		elif state in negativeStates:
+			# Translators: Indicates that a particular state on an object is negated.
```
Comment (dkager) [inline source/controlTypes.py]: "of an object"

**Review by dkager — CHANGES_REQUESTED** [review]
Looks pretty good.

**Comment by leonardder** [comment]
@dkager: Could you review my most recent commits?

**Review by dkager — APPROVED** [review]

**Comment by dkager** [comment]
As far as small changes bringing big joy go, this is still one of my favorites.

**Review by michaelDCurran — APPROVED** [review]

**Pull request merged**